### PR TITLE
Fix slasher disk usage leak.

### PR DESCRIPTION
**Pull request by nalepae**
**What type of PR is this?**
Bug fix

**What does this PR do? Why is it needed?**
**Please read commit by commit.**
Commits message explain how, and more importantly, why, we implement some changes.

This pull request fixes the slasher disk usage leak.

- The first part of this pull request solves the issue by encoding slot/epoch as a big-endian instead of little-endian in buckets keys.
- The second part of this pull request implements a database migration to ensure the retro-compatibility fo the database.

**Which issues(s) does this PR fix?**
- https://github.com/prysmaticlabs/prysm/issues/14142
- It may also fix https://github.com/prysmaticlabs/prysm/issues/13658 (but my intuition is there is an other bug to solve to fully fix this ticket)

### CODE REVIEW

File: beacon-chain/slasher/service.go [hunk 1]
```diff
@@ -121,6 +121,25 @@ func (s *Service) run() {
...
+	maxPruningEpoch := primitives.Epoch(0)
+	if headEpoch >= s.params.historyLength {
+		maxPruningEpoch = headEpoch - s.params.historyLength
+	}
+
+	// For database performance reasons, database read/write operations
+	// are chunked into batches of maximum `batchSize` elements.
+	const migrationBatchSize = 100_000
+
+	err = s.serviceCfg.Database.Migrate(s.ctx, headEpoch, maxPruningEpoch, migrationBatchSize)
```
Comment (prestonvanloon) [inline beacon-chain/slasher/service.go:137]: Does this perform the migrate every time the client starts?

Comment (nalepae) [reply beacon-chain/slasher/service.go:137]: Yes.

File: beacon-chain/slasher/service.go [hunk 2]
```diff
@@ -121,6 +121,25 @@ func (s *Service) run() {
...
+	if err != nil {
+		log.WithError(err).Error("Failed to migrate slasher database")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (nalepae) [reply beacon-chain/slasher/service.go:141]: Yes.

File: beacon-chain/slasher/service.go (hunk 1 above)
Comment (nalepae) [reply beacon-chain/slasher/service.go:137]: Yes, but if there is nothing to migrate (no epoch/slot higher than the current epoch/slot), then this functions exits immediately. Finding the highest epoch/slot in the DB is `O(1)`.

Comment (prestonvanloon) [reply beacon-chain/slasher/service.go:137]: OK, thanks

File: beacon-chain/db/slasherkv/migrate.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+// Migrate , its corresponding usage and tests can be totally removed once Electra is on mainnet.
+// Previously, the first 8 bytes of keys of `attestation-data-roots` and `proposal-records` buckets
+// were stored as little-endian respectively epoch and slots. It was the source of
+// https://github.com/prysmaticlabs/prysm/issues/14142 and potentially
+// https://github.com/prysmaticlabs/prysm/issues/13658.
+// To solve this (or these) issue(s), we decided to store the first 8 bytes of keys as big-endian.
+// See https://github.com/prysmaticlabs/prysm/pull/14151.
+// However, not to break the backward compatibility, we need to migrate the existing data.
+// The strategy is quite simple: If, for these bucket keys in the store, we detect
+// a slot (resp. epoch) higher, than the curreet slot (resp. epoch), then we consider that the data
```
Comment (prestonvanloon) [inline beacon-chain/db/slasherkv/migrate.go:24]:
```suggestion
// a slot (resp. epoch) higher, than the current slot (resp. epoch), then we consider that the data
```

**Review by prestonvanloon — APPROVED** [review]

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**